# Week03 Section 

## Data Wrangling with Pandas

Wenjun Sun

## What did we learn this week?

Let's go through each topic, and please raise your hand (using Zoom's raise-hand button) if you have questions about this topic.


- Series and DataFrame
- Load data 
- Exploratory Data Analysis (EDA): `info`, `head`, `shape`, `describe`, `unique`, `ProfileReport`
- Method Chaining
- Select columns and rows of a dataframe
- Apply functions to a dataframe
- `groupby` and summarize
- Index
- Reshape a dataframe
- Visualize
- Other topics

## Topics we will cover today 

I planned some topics for this section. But I priorizie answering all of your questions, so we are very likely not be able to cover all the materials, but I'll make sure to link to a tuturial that I found helpful for each topic and you can find this notebook at the `QMSS-G5072-2020/Wenjun_section_material` repository after our section. 

I would also really appreciate if you suggest topics. 

## Load and save data 

You should already know have to read in a dataset with `pd.read_..()`, for example `pd.read_csv()`. 

tutorial: https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html

In [2]:
import pandas as pd 

country = pd.read_csv('data/country.csv')
country.head(3) #oops, no column name?

Armenia           Asia
0         Algeria         Africa
1  U.S.S.R/Russia         Europe
2            U.S.  North America

In [14]:
pd.read_csv('data/country.csv', names=['country', 'continent']).head(3)

country continent
0         Armenia      Asia
1         Algeria    Africa
2  U.S.S.R/Russia    Europe

In [17]:
pd.read_csv('data/country.csv', names=['country', 'continent'], index_col='country').head(3) # set a column as index

continent
country                 
Armenia             Asia
Algeria           Africa
U.S.S.R/Russia    Europe

Similarly, we can save a DataFrame as a CSV file with `.to_csv()`

In [23]:
country = pd.read_csv('data/country.csv', names=['country', 'continent']).sample(5) # take a sample
country

country      continent
16  Saudi Arabia           Asia
18        Mexico  North America
0        Armenia           Asia
21         Japan           Asia
12       Vietnam           Asia

In [20]:
country.to_csv('data/country_sample.csv')

## `.select_dtypes()`

Return a subset of the DataFrame’s columns based on the column dtypes.

tutorial: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.select_dtypes.html?highlight=select_dtypes

In [22]:
astronauts = pd.read_csv('data/astronauts.csv')
astronauts.head(3) 

name gender  birth     nationality military_civilian  \
0        Gagarin, Yuri   male   1934  U.S.S.R/Russia          military   
1       Titov, Gherman   male   1935  U.S.S.R/Russia          military   
2  Glenn, John H., Jr.   male   1921            U.S.          military   

   mission_number occupation  year_of_mission  mission_hours mission_title  
0               1      pilot             1961           1.77      Vostok 1  
1               1      pilot             1961          25.00      Vostok 2  
2               1      pilot             1962           5.00          MA-6

In [26]:
astronauts.select_dtypes('number').head(3)

birth  mission_number  year_of_mission  mission_hours
0   1934               1             1961           1.77
1   1935               1             1961          25.00
2   1921               1             1962           5.00

In [32]:
astronauts.select_dtypes(exclude='number').head(3)

name gender     nationality military_civilian occupation  \
0        Gagarin, Yuri   male  U.S.S.R/Russia          military      pilot   
1       Titov, Gherman   male  U.S.S.R/Russia          military      pilot   
2  Glenn, John H., Jr.   male            U.S.          military      pilot   

  mission_title  
0      Vostok 1  
1      Vostok 2  
2          MA-6

## track statistics over rows

`cumsum()`, `cummax()`, `cummin()`, `cumprod()`

tutorial: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.cumsum.html?highlight=cumsum#pandas.DataFrame.cumsum

In [34]:
s = pd.Series([2, 3, 5, -1, 0])
s

0    2
1    3
2    5
3   -1
4    0
dtype: int64

In [35]:
s.cumsum()

0     2
1     5
2    10
3     9
4     9
dtype: int64

In [37]:
s.cumprod()

0     2
1     6
2    30
3   -30
4     0
dtype: int64

`cumcount()` only works with grouped data

In [41]:
df = pd.DataFrame([['a'], ['a'], ['a'], ['b'], ['b'], ['a']],
                  columns=['col'])
df

col
0   a
1   a
2   a
3   b
4   b
5   a

In [42]:
df.groupby('col').cumcount()

0    0
1    1
2    2
3    0
4    1
5    3
dtype: int64

## `to_markdown()`

Print DataFrame in Markdown-friendly format

tutorial: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_markdown.html?highlight=markdown#pandas.DataFrame.to_markdown

In [48]:
astronauts.head(5)

name gender  birth     nationality military_civilian  \
0        Gagarin, Yuri   male   1934  U.S.S.R/Russia          military   
1       Titov, Gherman   male   1935  U.S.S.R/Russia          military   
2  Glenn, John H., Jr.   male   1921            U.S.          military   
3  Glenn, John H., Jr.   male   1921            U.S.          military   
4  Carpenter, M. Scott   male   1925            U.S.          military   

   mission_number occupation  year_of_mission  mission_hours    mission_title  
0               1      pilot             1961           1.77         Vostok 1  
1               1      pilot             1961          25.00         Vostok 2  
2               1      pilot             1962           5.00             MA-6  
3               2      pilot             1998         213.00           STS-95  
4               1      pilot             1962           5.00  Mercury-Atlas 7

In [49]:
print(astronauts.head(5).to_markdown())

|    | name                | gender   |   birth | nationality    | military_civilian   |   mission_number | occupation   |   year_of_mission |   mission_hours | mission_title   |
|---:|:--------------------|:---------|--------:|:---------------|:--------------------|-----------------:|:-------------|------------------:|----------------:|:----------------|
|  0 | Gagarin, Yuri       | male     |    1934 | U.S.S.R/Russia | military            |                1 | pilot        |              1961 |            1.77 | Vostok 1        |
|  1 | Titov, Gherman      | male     |    1935 | U.S.S.R/Russia | military            |                1 | pilot        |              1961 |           25    | Vostok 2        |
|  2 | Glenn, John H., Jr. | male     |    1921 | U.S.           | military            |                1 | pilot        |              1962 |            5    | MA-6            |
|  3 | Glenn, John H., Jr. | male     |    1921 | U.S.           | military            |                2

This is a markdown table:

|    | name                | gender   |   birth | nationality    | military_civilian   |   mission_number | occupation   |   year_of_mission |   mission_hours | mission_title   |
|---:|:--------------------|:---------|--------:|:---------------|:--------------------|-----------------:|:-------------|------------------:|----------------:|:----------------|
|  0 | Gagarin, Yuri       | male     |    1934 | U.S.S.R/Russia | military            |                1 | pilot        |              1961 |            1.77 | Vostok 1        |
|  1 | Titov, Gherman      | male     |    1935 | U.S.S.R/Russia | military            |                1 | pilot        |              1961 |           25    | Vostok 2        |
|  2 | Glenn, John H., Jr. | male     |    1921 | U.S.           | military            |                1 | pilot        |              1962 |            5    | MA-6            |
|  3 | Glenn, John H., Jr. | male     |    1921 | U.S.           | military            |                2 | pilot        |              1998 |          213    | STS-95          |
|  4 | Carpenter, M. Scott | male     |    1925 | U.S.           | military            |                1 | pilot        |              1962 |            5    | Mercury-Atlas 7 |

## show a progress bar

tqdm is a very useful package that helps predict when theses operations will finish executing

tutorial: https://github.com/tqdm/tqdm

In [1]:
import numpy as np
from tqdm import tqdm, tqdm_notebook # install package tqdm first if necessary

In [9]:
df = pd.DataFrame(np.random.randint(0, 100, (100000, 1000)))

# instantiate
tqdm.pandas()

# Now you can use `progress_apply` instead of `apply`
df.progress_apply(lambda x: x**2)

/usr/local/lib/python3.8/site-packages/tqdm/std.py:706: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  from pandas import Panel
100%|██████████| 1000/1000 [00:01<00:00, 632.97it/s]


0     1     2     3     4     5     6     7     8     9    ...   990  \
0      6724  5929   324   169  1156   441  1764    36  4356  3721  ...  1369   
1      4225  9409  4225    81     1  3249  3969   576  3844  7056  ...   484   
2      6724  2209  3249  2601     1   841     0  4356   289  7225  ...  2209   
3      8464  3025  1936  8281   225   324  6889  3721  3025   676  ...   961   
4       225  1936  7396  9025   441   256  9801  5776  2209  7921  ...  8649   
...     ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
99995  1225    36  3364  6561  9216  3249  7396  5929   400  5329  ...   784   
99996  1024  4624  3721  4761  8281  3844  5329    81  5041  5476  ...   900   
99997    36   289     1  5329  9409  7225  1369   625  7569  2025  ...  9409   
99998    16  7056   784  5776  4761  8836   900  3249  3969  3364  ...  2304   
99999  1296  6084   289  7056  9025  1024    25   256   100  2601  ...  7569   

        991   992   993   994   995   996   997   998   999  
0       100  4761  3969  3136   784  9801   900    36    49  
1      6889  4624  2809  6400   289  7056   961   121  7056  
2      4356  5041   841    16  3481     0  1849  2116  1681  
3      2209  3721  7396  2500  2209  2025  5476  1089  5329  
4      2209  5476  7056   441  3249    16  2601  2809   225  
...     ...   ...   ...   ...   ...   ...   ...   ...   ...  
99995  8649  5776  5184  2601  6400   144  3249  9025  1936  
99996  9409  7569     1  8649  5776   441  6400  2209  7921  
99997  1681  6084  1849   676  7396  5776  6400  1521  5776  
99998  9409   169  7744   841  2601  7396  5776   196    16  
99999   900  8649  1089  1849  2401  5476    81  8464  7744  

[100000 rows x 1000 columns]

## Other topics I think useful

- [MultiIndex / advanced indexing](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html)
- [How to handle time series data with ease?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/09_timeseries.html)